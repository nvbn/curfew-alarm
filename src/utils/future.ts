/**
 * Abstraction around value that might not be yet available,
 * can be used with hooks when the values is produce asynchronously.
 * */

export const FUTURE_NOT_READY = "__FUTURE_NOT_READY__";

export type Future<T> = T | typeof FUTURE_NOT_READY;

export const isReady = <T>(val: Future<T>): val is T =>
  val !== FUTURE_NOT_READY;

export const futureMap = <T, Q>(v: Future<T>, fn: (v: T) => Q): Future<Q> =>
  isReady(v) ? fn(v) : v;
