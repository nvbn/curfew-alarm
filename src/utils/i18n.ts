import i18n, { TranslateOptions } from "i18n-js";

import { DEFAULT } from "../translations";

type Copy = keyof typeof DEFAULT;

type Scope = Copy | Copy[];

/**
 * i18n-js.t with typescript tags validation.
 */
const t = (scope: Scope, options?: TranslateOptions): string =>
  i18n.t(scope, options);

export default {
  ...i18n,
  t,
};
