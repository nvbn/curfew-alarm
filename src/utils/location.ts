import { INetwork, NETWORK_TYPE_WIFI } from "../dependencies/Network";

/**
 * When a user connected to wifi treated as they are at home.
 */
export const isAtHome = async (network: INetwork): Promise<boolean> => {
  const status = await network.getNetworkStateAsync();

  return status.type === NETWORK_TYPE_WIFI;
};
