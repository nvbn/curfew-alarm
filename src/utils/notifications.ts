import IConstants from "../dependencies/IConstants";
import INotifications, {
  NOTIFICATION_PERMISSIONS_DENIED,
  NOTIFICATIONS_PERMISSIONS_GRANTED,
  NOTIFICATIONS_PERMISSIONS_UNDETERMINED,
} from "../dependencies/INotifications";
import IPlatform, { PLATFORM_OS_ANDROID } from "../dependencies/IPlatform";

/**
 * Schedules local push notification.
 */
export const sendNotification = (
  notifications: INotifications,
  title: string,
  body: string,
): Promise<string> =>
  notifications.scheduleNotificationAsync({
    content: {
      title,
      body,
    },
    trigger: null,
  });

export const REGISTER_NOT_DEVICE = "REGISTER_NOT_DEVICE";

export const REGISTER_DENIED = "REGISTER_DECLINED";

export const REGISTER_OK = "REGISTER_OK";

type RegisterStatus = {
  status:
    | typeof REGISTER_NOT_DEVICE
    | typeof REGISTER_DENIED
    | typeof REGISTER_OK;
  token?: string;
};

/**
 * Registers app as push notifications sender / handler.
 */
export const registerForPushNotifications = async (
  { isDevice }: IConstants,
  { OS }: IPlatform,
  notifications: INotifications,
  reRequestPermissions: boolean,
): Promise<RegisterStatus> => {
  if (!isDevice) {
    return { status: REGISTER_NOT_DEVICE };
  }

  let status = (await notifications.getPermissionsAsync()).status;

  if (
    (reRequestPermissions && status === NOTIFICATION_PERMISSIONS_DENIED) ||
    status === NOTIFICATIONS_PERMISSIONS_UNDETERMINED
  ) {
    status = (await notifications.requestPermissionsAsync()).status;
  }

  if (status !== NOTIFICATIONS_PERMISSIONS_GRANTED) {
    return { status: REGISTER_DENIED };
  }

  const token = (await notifications.getExpoPushTokenAsync()).data;

  if (OS === PLATFORM_OS_ANDROID) {
    await notifications.setNotificationChannelAsync("default", {
      name: "default",
      importance: 5, // default
    });
  }

  return {
    status: REGISTER_OK,
    token,
  };
};
