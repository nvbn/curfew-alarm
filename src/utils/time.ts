import { isValidNumber } from "./numbers";

/**
 * Internal abstraction for storing time, it was just more convenient to have it.
 */
export type Time = {
  hour: number;
  minute: number;
};

/**
 * Time to minute from the start of the day.
 */
export const toMinutes = ({ hour, minute }: Time): number => hour * 60 + minute;

/**
 * Converts js Date to internal Time.
 */
export const dateToTime = (date: Date): Time => ({
  hour: date.getHours(),
  minute: date.getMinutes(),
});

const WE_DONT_CARE_ABOUT_THE_ACTUAL_DATE = "2021-01-10T00:00:00";

/**
 * Converts internal Time to js Date.
 */
export const timeToDate = ({ hour, minute }: Time): Date => {
  const date = new Date(WE_DONT_CARE_ABOUT_THE_ACTUAL_DATE);
  date.setHours(hour);
  date.setMinutes(minute);

  return date;
};

/**
 * Formats internal time in 24h form.
 */
export const formatTime = ({ hour, minute }: Time): string =>
  `${hour.toString().padStart(2, "0")}:${minute.toString().padStart(2, "0")}`;

/**
 * Returns true if passed object is a valid Time.
 */
// eslint-disable-next-line
export const isValidTime = (value: any): value is Time => {
  if (!value) {
    return false;
  }

  if (!isValidNumber(value.hour) || value.hour > 23 || value.hour < 0) {
    return false;
  }

  return !(
    !isValidNumber(value.minute) ||
    value.minute > 59 ||
    value.minute < 0
  );
};
