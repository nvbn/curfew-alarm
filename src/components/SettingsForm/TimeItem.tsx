import RNDateTimePicker, {
  Event,
} from "@react-native-community/datetimepicker";
import React, { useCallback, useState } from "react";
import { Text, TouchableOpacity, View } from "react-native";

import { PLATFORM_OS_ANDROID, PlatformOS } from "../../dependencies/IPlatform";
import { Future, futureMap, isReady } from "../../utils/future";
import { dateToTime, formatTime, Time, timeToDate } from "../../utils/time";
import styles from "./styles";
import { TimeItemProps } from "./types";

type TimeItemIOSProps = {
  id: string;
  title: string;
  valueAsDate: Future<Date>;
  onChange: (event: Event, newValue?: Date) => void;
};

const TimeItemIOS = ({
  id,
  title,
  valueAsDate,
  onChange,
}: TimeItemIOSProps): JSX.Element => (
  <View style={styles.item}>
    <Text style={styles.title}>{title}</Text>
    {isReady(valueAsDate) && (
      <RNDateTimePicker
        testID={`time-item-ios-time-picker-${id}`}
        mode="time"
        value={valueAsDate}
        onChange={onChange}
      />
    )}
  </View>
);

type TimeItemAndroidProps = {
  id: string;
  title: string;
  value: Future<Time>;
  valueAsDate: Future<Date>;
  onChange: (event: Event, newValue?: Date) => void;
};

const TimeItemAndroid = ({
  id,
  title,
  value,
  valueAsDate,
  onChange,
}: TimeItemAndroidProps): JSX.Element => {
  const [showPicker, setShowPicker] = useState(false);

  const onChangeWithDismiss = useCallback(
    (event: Event, newValue?: Date) => {
      setShowPicker(false);
      onChange(event, newValue);
    },
    [setShowPicker, onChange],
  );

  return (
    <TouchableOpacity
      testID={`time-item-android-touchable-${id}`}
      style={styles.item}
      onPress={() => {
        setShowPicker(true);
      }}
    >
      <Text style={styles.title}>{title}</Text>
      {isReady(value) && isReady(valueAsDate) && (
        <>
          <Text style={styles.input}>{formatTime(value)}</Text>
          {showPicker && (
            <RNDateTimePicker
              testID={`time-item-android-time-picker-${id}`}
              mode="time"
              value={valueAsDate}
              onChange={onChangeWithDismiss}
            />
          )}
        </>
      )}
    </TouchableOpacity>
  );
};

type Props = TimeItemProps & {
  os: PlatformOS;
};

/**
 * Allows to modify settings value that store time.
 */
const TimeItem = ({ id, title, value, onChange, os }: Props): JSX.Element => {
  const valueAsDate = futureMap(value, timeToDate);

  const onChangeAsTime = useCallback(
    (_: Event, newValue?: Date): void =>
      newValue && onChange(dateToTime(newValue)),
    [onChange],
  );

  return os === PLATFORM_OS_ANDROID ? (
    <TimeItemAndroid
      id={id}
      title={title}
      value={value}
      valueAsDate={valueAsDate}
      onChange={onChangeAsTime}
    />
  ) : (
    <TimeItemIOS
      id={id}
      title={title}
      valueAsDate={valueAsDate}
      onChange={onChangeAsTime}
    />
  );
};

export default TimeItem;
