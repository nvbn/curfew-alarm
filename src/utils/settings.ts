import IPersistentStorage from "../dependencies/IPersistentStorage";
import { isValidNumber } from "./numbers";
import { isValidTime, Time } from "./time";

export type Settings = {
  curfewStart: Time;
  curfewEnd: Time;

  minutesToGoHome: number;
};

// exported only for tests
export const DEFAULT_SETTINGS: Settings = {
  curfewStart: { hour: 21, minute: 0 },
  curfewEnd: { hour: 4, minute: 30 },
  minutesToGoHome: 30,
};

// exported only for tests
export const STORAGE_KEY = "SETTINGS";

const parseSettings = (serialised: string | null): Settings => {
  let settings = DEFAULT_SETTINGS;
  if (serialised !== null) {
    try {
      settings = { ...settings, ...JSON.parse(serialised) };
    } catch (e) {
      console.warn("unable to decode settings", e);
    }
  }

  if (!isValidTime(settings.curfewStart)) {
    console.warn("incorrect value in curfewStart", settings.curfewStart);
    settings.curfewStart = DEFAULT_SETTINGS.curfewStart;
  }

  if (!isValidTime(settings.curfewEnd)) {
    console.warn("incorrect value in curfewEnd", settings.curfewEnd);
    settings.curfewEnd = DEFAULT_SETTINGS.curfewEnd;
  }

  if (
    !isValidNumber(settings.minutesToGoHome) ||
    settings.minutesToGoHome < 0
  ) {
    console.warn(
      "incorrect value in minutesToGoHome",
      settings.minutesToGoHome,
    );
    settings.minutesToGoHome = DEFAULT_SETTINGS.minutesToGoHome;
  }

  return settings;
};

/**
 * Reads settings from persistent storage, returns default settings when
 * it's not possible, and replaces invalid keys with default values.
 */
export const getSettings = async (
  storage: IPersistentStorage,
): Promise<Settings> => {
  try {
    const serialised = await storage.getItem(STORAGE_KEY);

    return parseSettings(serialised);
  } catch (e) {
    console.warn("unable to read settings", e);

    return DEFAULT_SETTINGS;
  }
};

/**
 * Saves settings in persisten storage.
 */
export const updateSettings = (
  storage: IPersistentStorage,
  settings: Settings,
): Promise<void> => storage.setItem(STORAGE_KEY, JSON.stringify(settings));
