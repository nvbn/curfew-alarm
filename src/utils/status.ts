import { Time, toMinutes } from "./time";

export const STATUS_FINE = "FINE";

export const STATUS_GO_HOME_WHEN_TIME_TO_GO_HOME =
  "GO_HOME_WHEN_TIME_TO_GO_HOME";

export const STATUS_STAY_AT_HOME_WHEN_TIME_TO_GO_HOME =
  "STAY_AT_HOME_WHEN_TIME_TO_GO_HOME";

export const STATUS_STAY_AT_HOME_WHEN_CURFEW = "STAY_AT_HOME_WHEN_CURFEW";

export const STATUS_GO_HOME_WHEN_CURFEW = "GO_HOME_WHEN_CURFEW";

export type Status =
  | typeof STATUS_FINE
  | typeof STATUS_GO_HOME_WHEN_TIME_TO_GO_HOME
  | typeof STATUS_STAY_AT_HOME_WHEN_TIME_TO_GO_HOME
  | typeof STATUS_STAY_AT_HOME_WHEN_CURFEW
  | typeof STATUS_GO_HOME_WHEN_CURFEW;

const MAX_MINUTE = 24 * 60;

/**
 * Determines users status / what a user needs to do at the moment
 * based on time and configurations.
 */
export const getStatus = (
  enabled: boolean,
  isAtHome: boolean,
  currentTime: Time,
  curfewStart: Time,
  curfewEnd: Time,
  minutesToGoHome: number,
): Status => {
  if (!enabled) {
    return STATUS_FINE;
  }

  const currentMinute = toMinutes(currentTime);
  const curfewStartMinute = toMinutes(curfewStart);
  const curfewEndMinute = toMinutes(curfewEnd);
  const goHomeStartMinute = curfewStartMinute - minutesToGoHome;

  if (
    currentMinute >= curfewStartMinute &&
    ((curfewEndMinute < curfewStartMinute && currentMinute <= MAX_MINUTE) ||
      currentMinute <= curfewEndMinute)
  ) {
    return isAtHome
      ? STATUS_STAY_AT_HOME_WHEN_CURFEW
      : STATUS_GO_HOME_WHEN_CURFEW;
  } else if (currentMinute >= goHomeStartMinute) {
    return isAtHome
      ? STATUS_STAY_AT_HOME_WHEN_TIME_TO_GO_HOME
      : STATUS_GO_HOME_WHEN_TIME_TO_GO_HOME;
  } else {
    return STATUS_FINE;
  }
};
