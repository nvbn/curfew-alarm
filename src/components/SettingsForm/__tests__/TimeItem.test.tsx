/* eslint-disable @typescript-eslint/no-empty-function, @typescript-eslint/no-explicit-any */
import RNDateTimePicker, {
  AndroidNativeProps,
  IOSNativeProps,
} from "@react-native-community/datetimepicker";
import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import { TextInput } from "react-native";

import {
  PLATFORM_OS_ANDROID,
  PLATFORM_OS_IOS,
} from "../../../dependencies/IPlatform";
import { FUTURE_NOT_READY } from "../../../utils/future";
import { timeToDate } from "../../../utils/time";
import TimeItem from "../TimeItem";

jest.mock("@react-native-community/datetimepicker", () => jest.fn());

(RNDateTimePicker as jest.Mock).mockImplementation(
  ({ testID, value, onChange }: IOSNativeProps | AndroidNativeProps) => (
    <TextInput
      testID={testID}
      value={value.toString()}
      onChangeText={(textValue) => onChange!({} as any, new Date(textValue))}
    />
  ),
);

describe("<TimeItem />", () => {
  for (const os of [PLATFORM_OS_ANDROID, PLATFORM_OS_IOS]) {
    test(`can be rendered on ${os}`, () => {
      const rendered = render(
        <TimeItem
          id="test"
          title="When it starts"
          type="time"
          value={{
            hour: 20,
            minute: 30,
          }}
          onChange={() => {}}
          os={os}
        />,
      );

      expect(rendered.toJSON()).toMatchSnapshot();
    });

    test(`can be rendered when not ready on ${os}`, () => {
      const rendered = render(
        <TimeItem
          id="test"
          title="When it starts"
          type="time"
          value={FUTURE_NOT_READY}
          onChange={() => {}}
          os={os}
        />,
      );

      expect(rendered.toJSON()).toMatchSnapshot();
    });
  }

  test("can change time on android", () => {
    const onChange = jest.fn();

    const { getByTestId } = render(
      <TimeItem
        id="test"
        title="When it starts"
        type="time"
        value={{
          hour: 20,
          minute: 30,
        }}
        onChange={onChange}
        os={PLATFORM_OS_ANDROID}
      />,
    );

    const touchable = getByTestId("time-item-android-touchable-test");
    fireEvent.press(touchable);

    const timePicker = getByTestId("time-item-android-time-picker-test");
    const timestamp = timeToDate({ hour: 19, minute: 45 });
    fireEvent.changeText(timePicker, timestamp); // ugly mock event name

    expect(onChange.mock.calls).toEqual([[{ hour: 19, minute: 45 }]]);
  });

  test("can change time on ios", () => {
    const onChange = jest.fn();

    const { getByTestId } = render(
      <TimeItem
        id="test"
        title="When it starts"
        type="time"
        value={{
          hour: 20,
          minute: 30,
        }}
        onChange={onChange}
        os={PLATFORM_OS_IOS}
      />,
    );

    const timePicker = getByTestId("time-item-ios-time-picker-test");
    const timestamp = timeToDate({ hour: 21, minute: 15 });
    fireEvent.changeText(timePicker, timestamp); // ugly mock event name

    expect(onChange.mock.calls).toEqual([[{ hour: 21, minute: 15 }]]);
  });
});
