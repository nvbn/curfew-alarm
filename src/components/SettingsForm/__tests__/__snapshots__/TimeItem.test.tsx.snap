// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimeItem /> can be rendered on android 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderBottomColor": "#fafafa",
      "borderBottomWidth": 1,
      "marginHorizontal": 16,
      "marginVertical": 8,
      "opacity": 1,
      "paddingHorizontal": 10,
      "paddingVertical": 20,
    }
  }
  testID="time-item-android-touchable-test"
>
  <Text
    style={
      Object {
        "fontSize": 24,
        "paddingBottom": 10,
      }
    }
  >
    When it starts
  </Text>
  <Text
    style={
      Object {
        "fontSize": 18,
      }
    }
  >
    20:30
  </Text>
</View>
`;

exports[`<TimeItem /> can be rendered on ios 1`] = `
<View
  style={
    Object {
      "borderBottomColor": "#fafafa",
      "borderBottomWidth": 1,
      "marginHorizontal": 16,
      "marginVertical": 8,
      "paddingHorizontal": 10,
      "paddingVertical": 20,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 24,
        "paddingBottom": 10,
      }
    }
  >
    When it starts
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    rejectResponderTermination={true}
    testID="time-item-ios-time-picker-test"
    underlineColorAndroid="transparent"
    value="Sun Jan 10 2021 20:30:00 GMT+0000 (Coordinated Universal Time)"
  />
</View>
`;

exports[`<TimeItem /> can be rendered when not ready on android 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderBottomColor": "#fafafa",
      "borderBottomWidth": 1,
      "marginHorizontal": 16,
      "marginVertical": 8,
      "opacity": 1,
      "paddingHorizontal": 10,
      "paddingVertical": 20,
    }
  }
  testID="time-item-android-touchable-test"
>
  <Text
    style={
      Object {
        "fontSize": 24,
        "paddingBottom": 10,
      }
    }
  >
    When it starts
  </Text>
</View>
`;

exports[`<TimeItem /> can be rendered when not ready on ios 1`] = `
<View
  style={
    Object {
      "borderBottomColor": "#fafafa",
      "borderBottomWidth": 1,
      "marginHorizontal": 16,
      "marginVertical": 8,
      "paddingHorizontal": 10,
      "paddingVertical": 20,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 24,
        "paddingBottom": 10,
      }
    }
  >
    When it starts
  </Text>
</View>
`;
